SF-3306 Add initial draft history UI

pmachapman opened the pull request:
This PR enables the draft history portion of the updated draft UI. This UI is hidden behind the feature flag "Preview new draft history interface", which should be enabled to access this UI. See the mockup in Balsamiq for details (linked to in the [JIRA ticket](https://jira.sil.org/browse/SF-3306)).

I have not implemented the full UI, but the initial draft history portion, with the ability to download and view previous drafts. Other JIRA issues (SF-3354, SF-3355, SF-3356, and SF-3357) have been created for the rest of the tasks to complete this entire feature.

Viewing/applying/downloading previous drafts will only work fully for drafts created with this PR (this includes the displaying of training books/languages selected).

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/web-xforge/3180)
<!-- Reviewable:end -->

pmachapman reviewed (commented):
> Dismissed @github-advanced-security[bot] from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3180)* status: 0 of 41 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (changes requested):
The front-end looks really good and is already relatively feature complete based on my observations. Thanks for the good work!

> Reviewed 22 of 39 files at r1, 20 of 20 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-:-OP7sbVl1RCzFlZyLniJ:b8pwra5)* status: all files reviewed, 24 unresolved discussions (waiting on @pmachapman)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.spec.ts` line 236 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7muXwDTOrGv8jzfWM:-OP7muXwDTOrGv8jzfWN:bl1mycn) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.spec.ts#L236)):*
> ```XML
>   describe('getStatus', () => {
>     it('should handle known build state strings', () => {
>       expect(component.getStatus(BuildStates.Active)).not.toBeUndefined();
> ```

Nit: Instead of the double negative, this can just be `toBeDefined()` here and below.



_Code quote:_
```XML
      expect(component.getStatus(BuildStates.Active)).not.toBeUndefined();
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.ts` line 317 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7ZWmb6oByGKaTw0LJ:-OP7ZWmb6oByGKaTw0LK:b-4cxroz) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.ts#L317)):*
> ```XML
> 
>       // Build the list of book numbers, first checking the build, then the project document if that is null
>       const books: number[] =
> ```

You will probably want to get a Set for the books, otherwise there is the chance of duplicate books when we do implement multiple translation sources.



_Code quote:_
```XML
const books: number[] =
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.ts` line 28 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7eAHd5YQLmr-lCcOP:-OP7eAHd5YQLmr-lCcOQ:b-s5eatt) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.ts#L28)):*
> ```XML
>     private readonly transloco: TranslocoService
>   ) {
>     this.activatedProject.projectId$.pipe(filterNullish(), take(1)).subscribe(projectId => {
> ```

You'll want to also pipe it through `quietTakeUntilDestroyed(destroyRef)`. I have seen articles explaining why take(1) is not guaranteed to cause the subscript to complete if the first value never gets emitted from the observable and the user moves off the component.



_Code quote:_
```XML
    this.activatedProject.projectId$.pipe(filterNullish(), take(1)).subscribe(projectId => {
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.ts` line 52 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7evwMErKGROV9_cQz:-OP7evwMErKGROV9_cR-:bb752ax) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.ts#L52)):*
> ```XML
>         return this.transloco.translate('draft_history_list.draft_faulted');
>       default:
>         // The latest build must abe a build that has finished
> ```

Typo



_Code quote:_
```XML
// The latest build must abe a build that has finished
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.spec.ts` line 171 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7_LfU88RiKApcb3aH:-OP7_LfU88RiKApcb3aI:b-kq1cfw) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.spec.ts#L171)):*
> ```XML
>       // SUT
>       service.getBuildHistory(projectId).subscribe(result => {
>         expect(result).toEqual(undefined);
> ```

Nit: You could use `.toBeUndefined()` here and below.



_Code quote:_
```XML
        expect(result).toEqual(undefined);
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.spec.ts` line 620 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7aI6j7i3kA6hVb-B0:-OP7aI6j7i3kA6hVb-B1:b-khi6dv) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.spec.ts#L620)):*
> ```XML
>       // SUT
>       service.getGeneratedDraftHistory(projectId, book, chapter).subscribe(result => {
>         expect(result).toEqual(undefined);
> ```

Nit: This could be `toBeUndefined()` here and below.



_Suggestion:_
```XML
        expect(result).toEqual(undefined)
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 274 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7rGBJAYDvzYrAWMKL:-OP7rGBJAYDvzYrAWMKM:bj8xbu) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L274)):*
> ```JSON
>     "draft_unknown": "Unknown",
>     "hide_model_training_data": "Hide model training data",
>     "show_model_training_data": "Show model training data",
> ```

Nit: We should name this model training books to avoid confusion with training data files.



_Code quote:_
```JSON
    "hide_model_training_data": "Hide model training data",
    "show_model_training_data": "Show model training data",
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 280 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7rS6o-W2ID74daEcf:-OP7rS6o-W2ID74daEcg:b9t6892) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L280)):*
> ```JSON
>     "draft_canceled": "Your draft was canceled",
>     "draft_completed": "Your draft is ready",
>     "draft_faulted": "The draft failed",
> ```

We should be consistent and choose "your" or "the" for these strings.



_Code quote:_
```JSON
    "draft_canceled": "Your draft was canceled",
    "draft_completed": "Your draft is ready",
    "draft_faulted": "The draft failed",
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html` line 3 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7hJwKDFdt0sAsObG-:-OP7hJwKDFdt0sAsObG0:by02c8) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html#L3)):*
> ```HTML
> <ng-container *transloco="let t; read: 'draft_history_entry'">
>   @if (entry != null) {
>     <div class="draft-entry-heading" (click)="headerClicked()" [class.expandable]="!forceDetailsOpen">
> ```

Material does have an expansion panel component that we are already using on the draft generation page. Is there a reason for not using it?



_Code quote:_
```HTML
<div class="draft-entry-heading" (click)="headerClicked()" [class.expandable]="!forceDetailsOpen">
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html` line 9 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7jaxYEh6hkBFXYEkT:-OP7jaxZ6dGbgQt8NZXP:b-j4y8jd) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html#L9)):*
> ```HTML
>       </div>
>       <span class="spacer"></span>
>       <span class="status" [style.color]="getStatus(entry.state).color">
> ```

You will want to add the colour as a class to this element, and then in the css you can specify what the colour should be if the element contains the class. I think that is better than specifically setting the color style. It adds a layer so that the colour is not as tightly coupled to status.



_Code quote:_
```HTML
      <span class="status" [style.color]="getStatus(entry.state).color">
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html` line 19 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7hyh16FRKJgDB5Zhb:-OP7hyh16FRKJgDB5Zhc:b-rpf5ba) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html#L19)):*
> ```HTML
>     @if (detailsOpen && hasDetails) {
>       <div class="draft-entry-body">
>         <ng-container *ngIf="canDownloadBuild">
> ```

We should probably avoid mixing the structural directives `*ngIf` and use the control flow version here and throughout this file.



_Code quote:_
```HTML
        <ng-container *ngIf="canDownloadBuild">
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html` line 29 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7iKEH0WDpkuxy8DsD:-OP7iKEH0WDpkuxy8DsE:b-s1pa7p) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html#L29)):*
> ```HTML
>           <p *ngIf="canDownloadBuild">
>             <a
>               href="javascript:;"
> ```

What is this link doing?



_Code quote:_
```HTML
              href="javascript:;"
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html` line 37 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7ibL-9EKAOmJAkWkN:-OP7ibL-9EKAOmJAkWkO:b-wdqujp) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html#L37)):*
> ```HTML
>             </a>
>           </p>
>           @if (trainingDataOpen) {
> ```

We might want to call this training books so that we don't get this confused with training data files that is in addition to training books.



_Code quote:_
```HTML
          </p>
          @if (trainingDataOpen) {
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.spec.ts` line 119 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7ft918ThlYHQ7Ugte:-OP7ft918ThlYHQ7Ugtf:b-gi4hic) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.spec.ts#L119)):*
> ```XML
>     constructor(buildHistory: BuildDto[] | undefined) {
>       when(mockedActivatedProjectService.projectId$).thenReturn(of('project01'));
>       when(mockedActivatedProjectService.changes$).thenReturn(of(undefined));
> ```

Nit: I don't think this line is necessary.



_Code quote:_
```XML
      when(mockedActivatedProjectService.changes$).thenReturn(of(undefined));
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor-draft/editor-draft.component.ts` line 259 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7qea5-ND4usJGQCau:-OP7qea5-ND4usJGQCav:b-slup6n) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor-draft/editor-draft.component.ts#L259)):*
> ```XML
>   }
> 
>   findClosestRevision(date: Date, revisions: Revision[]): Revision {
> ```

This can be a private method.



_Code quote:_
```XML
 findClosestRevision(date: Date, revisions: Revision[]): Revision {
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.html` line 5 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7nQtB3AzwdgfcJSNi:-OP7nQtB3AzwdgfcJSNj:bjm2wzr) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.html#L5)):*
> ```HTML
>     <mat-button-toggle-group
>       class="draft-book-option"
>       [disabled]="book.chaptersWithDrafts.length === 0"
> ```

I wouldn't expect this to be possible since the book wouldn't have been selected if it didn't have an chapters with a draft. Is there something I am not understanding?



_Code quote:_
```HTML
      [disabled]="book.chaptersWithDrafts.length === 0"
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor-draft/editor-draft.component.html` line 21 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7pK4eCncIqGan9ga_:-OP7pK4eCncIqGan9gaa:bjraa11) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor-draft/editor-draft.component.html#L21)):*
> ```HTML
>   @if (isDraftReady) {
>     <div class="toolbar">
>       <mat-form-field class="draft-select" appearance="outline">
> ```

This is a really neat addition. It works so seamlessly!



_Code quote:_
```HTML
      <mat-form-field class="draft-select" appearance="outline">
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-handling.service.ts` line 165 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7ba5sEbkAsBJ0knHn:-OP7ba5sEbkAsBJ0knHo:b5zfizr) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-handling.service.ts#L165)):*
> ```XML
>               // The legacy format does not support a timestamp
>               if (err.status === 405 && timestamp == null) {
>                 return this.getDraft(textDocId, { isDraftLegacy: true, timestamp });
> ```

If the legacy does not support time stamp, then we should just omit that here.



_Code quote:_
```XML
    return this.getDraft(textDocId, { isDraftLegacy: true, timestamp });
```
___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 1496 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7sPKeCzNF2RpVoxmS:-OP7sPKeCzNF2RpVoxmT:blge0wd) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L1496)):*
> ```Smalltalk
>     }
> 
>     private static string GetTranslationId(
> ```

This should be `GetTranslationEngineId` to make it less ambiguous that it is looking for the engine.



_Code quote:_
```Smalltalk
    private static string GetTranslationId(
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.ts` line 72 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7nz7S2CjqacJNngJk:-OP7nz7S2CjqacJNngJl:bbfhqg2) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.ts#L72)):*
> ```XML
>       } else {
>         draftBooks = this.build.additionalInfo?.translationScriptureRanges
>           .flatMap(range => booksFromScriptureRange(range.scriptureRange))
> ```

This will work currently, but once there are multiple translation sources, this could lead to duplicate books.



_Code quote:_
```XML
          .flatMap(range => booksFromScriptureRange(range.scriptureRange))
```
___
*[`src/SIL.XForge.Scripture/Models/ServalBuildAdditionalInfo.cs` line 12 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7rkMR1h5fxoIpsLqO:-OP7rkMR1h5fxoIpsLqP:bxxtm3f) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/Models/ServalBuildAdditionalInfo.cs#L12)):*
> ```Smalltalk
>     public DateTimeOffset? DateFinished { get; init; }
>     public DateTimeOffset? DateRequested { get; set; }
>     public DateTimeOffset? DateGenerated { get; set; }
> ```

I wonder if these should allow a setter, or if `init` would make more sense. Oh, nevermind, I see the use case for allowing the setter now. Good.



_Code quote:_
```Smalltalk
    public DateTimeOffset? DateRequested { get; set; }
    public DateTimeOffset? DateGenerated { get; set; }
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts` line 68 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7kUIYD2pGk3OjvxoE:-OP7kUIYD2pGk3OjvxoF:b1jpex4) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts#L68)):*
> ```XML
>     this._trainingData = [];
> 
>     // Get the training data
> ```

Nit: We will probably want to call this training books to not get this confused with training data files.



_Code quote:_
```XML
    // Get the training data
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts` line 129 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7lR5hCroIpME_FAOv:-OP7lR5hCroIpME_FAOw:b9yx8zb) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts#L129)):*
> ```XML
>   get columnsToDisplay(): string[] {
>     return ['bookNames', 'source', 'target'];
>   }
> ```

Nit: This doesn't need to be a getter. It can just be a field variable.



_Code quote:_
```XML
  get columnsToDisplay(): string[] {
    return ['bookNames', 'source', 'target'];
  }
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts` line 152 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7lgGAEIGw7swd9EkY:-OP7lgGAEIGw7swd9EkZ:blnnauw) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts#L152)):*
> ```XML
>   get trainingData(): TrainingDataRow[] {
>     return this._trainingData;
>   }
> ```

Nit: This would be better if it was called training books



_Code quote:_
```XML
  private _trainingData: TrainingDataRow[] = [];
  get trainingData(): TrainingDataRow[] {
    return this._trainingData;
  }
```


<!-- Sent from Reviewable.io -->

pmachapman reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3180)* status: all files reviewed, 16 unresolved discussions (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.ts` line 317 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7ZWmb6oByGKaTw0LJ:-OPS1_EkETnFX-05CJSr:b-ydonir) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.ts#L317)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

You will probably want to get a Set for the books, otherwise there is the chance of duplicate books when we do implement multiple translation sources.
</blockquote></details>

Done. Good idea.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.spec.ts` line 171 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7_LfU88RiKApcb3aH:-OPS2Ev_7oZUyWHcxLQf:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.spec.ts#L171)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: You could use `.toBeUndefined()` here and below.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.spec.ts` line 620 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7aI6j7i3kA6hVb-B0:-OPS2FnI2jKN0NW4WlCu:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-generation.service.spec.ts#L620)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: This could be `toBeUndefined()` here and below.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-handling.service.ts` line 165 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7ba5sEbkAsBJ0knHn:-OPS2MFMCgEWTHGg-TzL:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-handling.service.ts#L165)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

If the legacy does not support time stamp, then we should just omit that here.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.ts` line 28 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7eAHd5YQLmr-lCcOP:-OPS8zkl8-_Rr2G_SIXy:b-ioxi27) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.ts#L28)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

You'll want to also pipe it through `quietTakeUntilDestroyed(destroyRef)`. I have seen articles explaining why take(1) is not guaranteed to cause the subscript to complete if the first value never gets emitted from the observable and the user moves off the component.
</blockquote></details>

Done. Thank you!
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.ts` line 52 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7evwMErKGROV9_cQz:-OPSAGYI4zLTLZ2vy41-:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.ts#L52)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Typo
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.spec.ts` line 119 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7ft918ThlYHQ7Ugte:-OPSAvGG5bLbncdm71hg:b-noos0p) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.spec.ts#L119)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: I don't think this line is necessary.
</blockquote></details>

It is required for `DraftPreviewBooksComponent`. I have added a comment stating this.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html` line 3 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7hJwKDFdt0sAsObG-:-OPSYtQ80LUxBvas7nTp:bgifio4) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html#L3)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Material does have an expansion panel component that we are already using on the draft generation page. Is there a reason for not using it?
</blockquote></details>

Done. I have rewritten this as a `mat-expansion-panel`.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html` line 9 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7jaxYEh6hkBFXYEkT:-OPShsTm0PJh8UPVu_m4:b8ejfk2) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html#L9)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

You will want to add the colour as a class to this element, and then in the css you can specify what the colour should be if the element contains the class. I think that is better than specifically setting the color style. It adds a layer so that the colour is not as tightly coupled to status.
</blockquote></details>

Done. I have added dark mode support, too.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html` line 19 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7hyh16FRKJgDB5Zhb:-OPSiPs891ZBJK6CnMxS:b-yz6sbd) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html#L19)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

We should probably avoid mixing the structural directives `*ngIf` and use the control flow version here and throughout this file.
</blockquote></details>

Done. Old habits die hard!
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html` line 29 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7iKEH0WDpkuxy8DsD:-OPSiUYMEmIhDg-yZnZg:b-f0jze9) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html#L29)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

What is this link doing?
</blockquote></details>

That is the same as `href="#"`, but doesn't trigger a navigate.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html` line 37 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7ibL-9EKAOmJAkWkN:-OPSl-q_AJ5QBNAuqer_:b-1tvbv2) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.html#L37)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

We might want to call this training books so that we don't get this confused with training data files that is in addition to training books.
</blockquote></details>

I have renamed this and similar variables to be training configuration rather than training data.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts` line 68 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7kUIYD2pGk3OjvxoE:-OPSlBnKAZe7_G7DjZGo:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts#L68)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: We will probably want to call this training books to not get this confused with training data files.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts` line 129 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7lR5hCroIpME_FAOv:-OPSlN5n5jHdIJdg3vqy:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts#L129)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: This doesn't need to be a getter. It can just be a field variable.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts` line 152 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7lgGAEIGw7swd9EkY:-OPSlTyJ1NMHeeU9c-5x:b-bvhybp) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.ts#L152)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: This would be better if it was called training books
</blockquote></details>

I have called it `trainingConfiguration`, as `trainingBooks` made it confusing when refactoring it.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.spec.ts` line 236 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7muXwDTOrGv8jzfWM:-OPSloFc5JplcufZfDyg:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-entry/draft-history-entry.component.spec.ts#L236)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: Instead of the double negative, this can just be `toBeDefined()` here and below.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.html` line 5 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7nQtB3AzwdgfcJSNi:-OPSmB-u916f_0e-QAKZ:bjikvg2) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.html#L5)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

I wouldn't expect this to be possible since the book wouldn't have been selected if it didn't have an chapters with a draft. Is there something I am not understanding?
</blockquote></details>

If this is showing the books for a historic draft, and the book has been removed or not included in the latest draft, this length can be zero.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.ts` line 72 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7nz7S2CjqacJNngJk:-OPSmXcn7q_9-pCIJhrC:b-de73j3) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.ts#L72)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

This will work currently, but once there are multiple translation sources, this could lead to duplicate books.
</blockquote></details>

Yes. I am not sure how we will support this in the UI. I think we will tackle this and other similar challenges when multiple translation sources support is added.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor-draft/editor-draft.component.ts` line 259 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7qea5-ND4usJGQCau:-OPSn1NiDLBDxsqkRCkE:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor-draft/editor-draft.component.ts#L259)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

This can be a private method.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 274 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7rGBJAYDvzYrAWMKL:-OPSn3yYDEkvfmPuVIRq:b8u1rih) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L274)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Nit: We should name this model training books to avoid confusion with training data files.
</blockquote></details>

I have renamed this to "training configuration"
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 280 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7rS6o-W2ID74daEcf:-OPSnIX0CgF-HnGqD5ID:b9yyat8) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L280)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

We should be consistent and choose "your" or "the" for these strings.
</blockquote></details>

I have renamed these to be "The" not "Your".
___
*[`src/SIL.XForge.Scripture/Services/MachineApiService.cs` line 1496 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7sPKeCzNF2RpVoxmS:-OPSnbqL9D5Z0MN4DfDk:bugs9as) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/Services/MachineApiService.cs#L1496)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

This should be `GetTranslationEngineId` to make it less ambiguous that it is looking for the engine.
</blockquote></details>

Done. Good spotting.


<!-- Sent from Reviewable.io -->

Nateowami commented:
(I'm sorry this is going to be barrage of comments, some of which are probably just uninformed, but I'm catching up from being completely away for a while). I got tripped up quite a bit by not running the webhook, and only towards the end figured out what was wrong.

@pmachapman Is it intentional that this select is not subject to the feature flag? It shows up all the time.

![Screenshot from 2025-05-05 12-16-11](https://github.com/user-attachments/assets/51ab81fe-990c-4844-bd88-e6fd8000b4ef)

Some kind of heading is needed here to separate the history from the technical options, otherwise they run together and the UI looks broken.

![Screenshot from 2025-05-05 12-17-38](https://github.com/user-attachments/assets/484935e6-282c-480d-a729-b27b1166c113)

I'm also not sure why I can't expand the older history items. I think there isn't sufficient data? Right now it feels like the UI is broken. Can we make it expand, but just state that exact information isn't available for older drafts?

The spacing on the left side of this button is way less than it should be.
![Screenshot from 2025-05-05 12-22-49](https://github.com/user-attachments/assets/5e436033-9c11-48ef-980f-b2b41a8edbc3)

pmachapman commented:
> Is it intentional that this select is not subject to the feature flag? It shows up all the time.

@Nateowami Yes. I figured that we could have it independent of the feature flag. That said, if you think it should be hidden behind the feature flag, I can do that.

> Some kind of heading is needed here to separate the history from the technical options, otherwise they run together and the UI looks broken.

I've added a heading to separate this (only Serval admins will see the bottom two panels and the heading)

> I'm also not sure why I can't expand the older history items. I think there isn't sufficient data? Right now it feels like the UI is broken. Can we make it expand, but just state that exact information isn't available for older drafts?

I will look into that. Maybe I will see if there is some data we can show, even if partial.

> The spacing on the left side of this button is way less than it should be.

I have fixed it, but it wasn't a pretty fix (I had to duplicate the button HTML).

Nateowami commented:
@pmachapman Yes, the dropdown should be subject to the feature flag. We had some discussions earlier today about the fact that just showing the date is confusing, and some potential improvements. It's going to be complicated enough that I think it would be far easier to just feature flag it until we figure that out.

Nateowami reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3180)* status: 21 of 43 files reviewed, 17 unresolved discussions (waiting on @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-download-button/draft-download-button.component.html` line 12 at r4](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OPXLgM100Hs2PO2anqS:-OPXLgM100Hs2PO2anqT:blpk4i8) ([raw file](https://github.com/sillsdev/web-xforge/blob/b2ef29666f3eefd42aedad1d30b786d8f0aee81e/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-download-button/draft-download-button.component.html#L12)):*
> ```HTML
>   >
>     @if (downloadProgress === 0) {
>       <mat-icon [ngClass]="flat ? '' : 'material-icons-outlined'">cloud_download</mat-icon>
> ```

It's no longer relevant given your latest changes, but this conditional class could have been written much more simply as `[class.material-icons-outlined]="flat"`. It's also possible to use object syntax: `[ngClass]="{'material-icons-outlined': flat}"`



_Code quote:_
```HTML
[ngClass]="flat ? '' : 'material-icons-outlined'"
```


<!-- Sent from Reviewable.io -->

pmachapman reviewed (commented):
> Yes, the dropdown should be subject to the feature flag. We had some discussions earlier today about the fact that just showing the date is confusing, and some potential improvements. It's going to be complicated enough that I think it would be far easier to just feature flag it until we figure that out.

@Nateowami Done.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-:-OPXPD8e3AtJbp0IQHAZ:b-ggrmr9)* status: 20 of 43 files reviewed, 17 unresolved discussions (waiting on @Nateowami and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-download-button/draft-download-button.component.html` line 12 at r4](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OPXLgM100Hs2PO2anqS:-OPXPGYiA4m4Di4ZYb6v:bszzn3u) ([raw file](https://github.com/sillsdev/web-xforge/blob/b2ef29666f3eefd42aedad1d30b786d8f0aee81e/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-download-button/draft-download-button.component.html#L12)):*
<details><summary><i>Previously, Nateowami wrote…</i></summary><blockquote>

It's no longer relevant given your latest changes, but this conditional class could have been written much more simply as `[class.material-icons-outlined]="flat"`. It's also possible to use object syntax: `[ngClass]="{'material-icons-outlined': flat}"`
</blockquote></details>

Done. Thanks for letting me know!


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (changes requested):
> Reviewed 6 of 19 files at r3, 15 of 16 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3180)* status: all files reviewed, 4 unresolved discussions (waiting on @Nateowami and @pmachapman)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.spec.ts` line 119 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7ft918ThlYHQ7Ugte:-OPYlCgZDkgu5E0C2uNW:bi7i1nz) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-history-list/draft-history-list.component.spec.ts#L119)):*
<details><summary><i>Previously, pmachapman (Peter Chapman) wrote…</i></summary><blockquote>

It is required for `DraftPreviewBooksComponent`. I have added a comment stating this.
</blockquote></details>

Excellent. thanks for the explanation.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.ts` line 72 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7nz7S2CjqacJNngJk:-OPYp0cz8Ep9t6OHTEze:btjpsqe) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.ts#L72)):*
<details><summary><i>Previously, pmachapman (Peter Chapman) wrote…</i></summary><blockquote>

Yes. I am not sure how we will support this in the UI. I think we will tackle this and other similar challenges when multiple translation sources support is added.
</blockquote></details>

Agreed, we don't need to solve all the problems now. Let's leave a TODO here to remind us this needs work.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 218 at r6](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OPYnexm9A_Ej9PagLLX:-OPYnexm9A_Ej9PagLLY:b-txzi2t) ([raw file](https://github.com/sillsdev/web-xforge/blob/99c83022699435fe55f7321e88701af16ad586e3/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L218)):*
> ```JSON
>     "preview_last_draft_header": "Preview last draft",
>     "report_problem": "Report problem",
>     "serval_administration": "Serval administration",
> ```

We don't need to localize this. We should update the places where we have this internationalized.



_Code quote:_
```JSON
    "serval_administration": "Serval administration",
```
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor-draft/editor-draft.component.html` line 21 at r6](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OPYnLbu2BxgfSA19kMc:-OPYnLbu2BxgfSA19kMd:b-x3gtbh) ([raw file](https://github.com/sillsdev/web-xforge/blob/99c83022699435fe55f7321e88701af16ad586e3/src/SIL.XForge.Scripture/ClientApp/src/app/translate/editor/editor-draft/editor-draft.component.html#L21)):*
> ```HTML
>   @if (isDraftReady) {
>     <div class="toolbar">
>       @if (featureFlags.newDraftHistory.enabled) {
> ```

Thanks for catching this!



_Code quote:_
```HTML
     @if (featureFlags.newDraftHistory.enabled) {
```


<!-- Sent from Reviewable.io -->

pmachapman reviewed (commented):
> I'm also not sure why I can't expand the older history items. I think there isn't sufficient data? Right now it feels like the UI is broken. Can we make it expand, but just state that exact information isn't available for older drafts?

@Nateowami More information should now show for older builds Let me know if this looks ok, especially the use of "Unknown" for the source project.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-:-OPbFv7dABkD9X9oJbYu:bhjyswy)* status: 35 of 43 files reviewed, 2 unresolved discussions (waiting on @Nateowami and @RaymondLuong3)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.ts` line 72 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OP7nz7S2CjqacJNngJk:-OPbEkTA7RZe0XnFRSn1:b-896fix) ([raw file](https://github.com/sillsdev/web-xforge/blob/9cfd1d6392da314d4126a2b59eeeae39ac7a4fcd/src/SIL.XForge.Scripture/ClientApp/src/app/translate/draft-generation/draft-preview-books/draft-preview-books.component.ts#L72)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

Agreed, we don't need to solve all the problems now. Let's leave a TODO here to remind us this needs work.
</blockquote></details>

Done.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 218 at r6](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OPYnexm9A_Ej9PagLLX:-OPbF48599xsUxwht-zt:b-p17tnb) ([raw file](https://github.com/sillsdev/web-xforge/blob/99c83022699435fe55f7321e88701af16ad586e3/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L218)):*
<details><summary><i>Previously, RaymondLuong3 (Raymond Luong) wrote…</i></summary><blockquote>

We don't need to localize this. We should update the places where we have this internationalized.
</blockquote></details>

Done. I have removed this string. It also occurs in the top right hand app menu - do you want me to remove it (and probably Hangfire Dashboard) from there too?


<!-- Sent from Reviewable.io -->

RaymondLuong3 reviewed (approved):
> Reviewed 21 of 21 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/3180)* status: all files reviewed, 1 unresolved discussion (waiting on @Nateowami)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json` line 218 at r6](https://reviewable.io/reviews/sillsdev/web-xforge/3180#-OPYnexm9A_Ej9PagLLX:-OPfcWfqB1KKk1kgA_rm:b-j5jbpe) ([raw file](https://github.com/sillsdev/web-xforge/blob/99c83022699435fe55f7321e88701af16ad586e3/src/SIL.XForge.Scripture/ClientApp/src/assets/i18n/non_checking_en.json#L218)):*
<details><summary><i>Previously, pmachapman (Peter Chapman) wrote…</i></summary><blockquote>

Done. I have removed this string. It also occurs in the top right hand app menu - do you want me to remove it (and probably Hangfire Dashboard) from there too?
</blockquote></details>

Oh that's interesting. I am happy to remove it or leave it as is. Your call.


<!-- Sent from Reviewable.io -->
